source/extensions/access_loggers/http_grpc/grpc_access_log_impl.cc

```diff
@@ -32,64 +32,85 @@ TLSProperties_TLSVersion tlsVersionStringToEnum(const std::string& tls_version)
+    parent_.stream_.reset();
   }
 }
 
-GrpcAccessLogStreamerImpl::ThreadLocalStreamer::ThreadLocalStreamer(
-    const SharedStateSharedPtr& shared_state)
-    : client_(shared_state->factory_->create()), shared_state_(shared_state) {}
+GrpcAccessLoggerImpl::GrpcAccessLoggerImpl(Grpc::RawAsyncClientPtr&& client,
+                                           const std::string& log_name,
+                                           const LocalInfo::LocalInfo& local_info)
+    : client_(std::move(client)), log_name_(log_name), local_info_(local_info) {}
 
-void GrpcAccessLogStreamerImpl::ThreadLocalStreamer::send(
-    envoy::service::accesslog::v2::StreamAccessLogsMessage& message, const std::string& log_name) {
-  auto stream_it = stream_map_.find(log_name);
-  if (stream_it == stream_map_.end()) {
-    stream_it = stream_map_.emplace(log_name, ThreadLocalStream(*this, log_name)).first;
+void GrpcAccessLoggerImpl::log(envoy::data::accesslog::v2::HTTPAccessLogEntry& entry) {
+  // TODO(euroelessar): Add batching and flushing.
+  envoy::service::accesslog::v2::StreamAccessLogsMessage message;
+  message.mutable_http_logs()->add_log_entry()->Swap(&entry);
+
+  if (stream_ == absl::nullopt) {
+    stream_.emplace(*this);
   }
 
-  auto& stream_entry = stream_it->second;
-  if (stream_entry.stream_ == nullptr) {
-    stream_entry.stream_ =
+  if (stream_->stream_ == nullptr) {
+    stream_->stream_ =
         client_->start(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
                            "envoy.service.accesslog.v2.AccessLogService.StreamAccessLogs"),
-                       stream_entry);
+                       *stream_);
 
     auto* identifier = message.mutable_identifier();
-    *identifier->mutable_node() = shared_state_->local_info_.node();
-    identifier->set_log_name(log_name);
+    *identifier->mutable_node() = local_info_.node();
+    identifier->set_log_name(log_name_);
   }
 
-  if (stream_entry.stream_ != nullptr) {
-    stream_entry.stream_->sendMessage(message, false);
+  if (stream_->stream_ != nullptr) {
+    stream_->stream_->sendMessage(message, false);
   } else {
     // Clear out the stream data due to stream creation failure.
-    stream_map_.erase(stream_it);
+    stream_.reset();
   }
 }
 
+GrpcAccessLoggerCacheImpl::GrpcAccessLoggerCacheImpl(Grpc::AsyncClientManager& async_client_manager,
+                                                     Stats::Scope& scope,
+                                                     ThreadLocal::SlotAllocator& tls,
+                                                     const LocalInfo::LocalInfo& local_info)
+    : async_client_manager_(async_client_manager), scope_(scope), tls_slot_(tls.allocateSlot()),
+      local_info_(local_info) {
+  tls_slot_->set([](Event::Dispatcher&) {
+    return ThreadLocal::ThreadLocalObjectSharedPtr{new ThreadLocalCache()};
+  });
+}
+
+GrpcAccessLoggerSharedPtr GrpcAccessLoggerCacheImpl::getOrCreateLogger(
```

Comment:
Technically, you will never delete loggers that are no longer used. In practice, I don't think this is a real issue, but it's something that I would document in a TODO somewhere.